lib/_http_parser.js

```diff
@@ -0,0 +1,907 @@
+      case STATE_BODY_CHUNKED_SIZE_IGNORE:
+        // We only reach this state once we receive a non-hex character on the
+        // chunk size line
+        while (offset < len) {
+          if (data[offset++] === LF) {
+            if (nbytes === 0) {
+              this._flags |= FLAG_TRAILING;
+              this._state = STATE_HEADER;
+              this._nbytes = null;
+              this.execute = this._executeHeader;
+              if (offset < len) {
+                var ret = this.execute(data.slice(offset));
+                if (typeof ret !== 'number')
+                  return ret;
+                return offset + ret;
+              }
+              return offset;
+            } else {
+              state = STATE_BODY_CHUNKED_BYTES;
+              break;
+            }
+          }
+        }
+        break;
+      default:
+        return this._setError('Unexpected parser state while reading chunks');
+    }
+  }
+
+  this._state = state;
+  this._nbytes = nbytes;
+
+  return len;
+};
+HTTPParser.prototype._executeBodyLiteral = function(data) {
+  var len = data.length;
+
+  if (len === 0)
+    return 0;
+
+  var nbytes = this._contentLen;
+  if (len >= nbytes) {
+    this.onBody(data, 0, nbytes);
+    this.onComplete && this.onComplete();
+    if ((this._flags & FLAG_SHOULD_KEEP_ALIVE) > 0) {
+      this.reinitialize(this.type);
+      if (len > nbytes) {
+        var ret = this.execute(data.slice(nbytes));
+        if (typeof ret !== 'number')
+          return ret;
+        return nbytes + ret;
+      }
+    } else {
+      this._state = STATE_COMPLETE;
+    }
+  } else {
+    this._contentLen -= len;
+    this.onBody(data, 0, len);
+  }
+  return len;
+};
+HTTPParser.prototype._executeBodyEOF = function(data) {
+  var len = data.length;
+
+  if (len === 0)
+    return 0;
+
+  this.onBody(data, 0, len);
+  return len;
+};
+HTTPParser.prototype._executeBodyIgnore = function(data) {
+  return 0;
+};
+HTTPParser.prototype._executeError = function(data) {
+  return this._err;
+};
+HTTPParser.prototype.execute = HTTPParser.prototype._executeStartLine;
+HTTPParser.prototype._shouldKeepAlive = function(httpMajor, httpMinor, flags,
+                                                 type, status) {
+  if (httpMajor > 0 && httpMinor > 0) {
+    if ((flags & FLAG_CONNECTION_CLOSE) > 0)
+      return false;
+  } else {
+    if ((flags & FLAG_CONNECTION_KEEP_ALIVE) === 0)
+      return false;
+  }
+  return !this._needsEOF(flags, type, status);
+};
+HTTPParser.prototype._needsEOF = function(flags, type, status) {
+  if (type === REQUEST)
+    return false;
+
+  // See RFC 2616 section 4.4
+  if (status === 204 ||                    // No Content
+      status === 304 ||                    // Not Modified
+      (status >= 100 && status < 200) ||   // 1xx e.g. Continue
+      (flags & FLAG_SKIPBODY) > 0) {       // response to a HEAD request
+    return false;
+  }
+
+  if ((flags & FLAG_CHUNKED) > 0 || this._contentLen !== null)
```

Comment:
The condition already yields a boolean, so you shouldn't need an if-statement (and potential branch misprediction).

```js
return (flags & FLAG_CHUNKED) > 0 || this._contentLen !== null;
```